### PR TITLE
JAMES-2555 Avoid failing on large mailboxes

### PR DESCRIPTION
**Pull request by chibenwa**

### CODE REVIEW

File: mailbox/tools/indexer/src/main/java/org/apache/mailbox/tools/indexer/ReIndexerImpl.java [hunk 1]
```diff
@@ -137,10 +136,10 @@ private void handleMailboxIndexingIterations(MailboxSession mailboxSession, Mail
         while (iterator.hasNext()) {
             MailboxMessage message = iterator.next();
             ImpactingMessageEvent impactingMessageEvent = findMostRelevant(mailboxRegistration.getImpactingEvents(message.getUid()));
-            if (impactingMessageEvent == null) {
```
Comment (mbaechler) [inline mailbox/tools/indexer/src/main/java/org/apache/mailbox/tools/indexer/ReIndexerImpl.java]: you don't check for impactingMessageEvent == null anymore ?

File: mailbox/tools/indexer/src/main/java/org/apache/mailbox/tools/indexer/ReIndexerImpl.java [hunk 2]
```diff
@@ -135,23 +134,27 @@ private void handleFullReindexingIterations(List<MailboxPath> mailboxPaths, Glob
...
         while (iterator.hasNext()) {
             MailboxMessage message = iterator.next();
-            ImpactingMessageEvent impactingMessageEvent = findMostRelevant(mailboxRegistration.getImpactingEvents(message.getUid()));
+            Optional<ImpactingMessageEvent> impactingMessageEvent = findMostRelevant(mailboxRegistration.getImpactingEvents(message.getUid()));
 
-            impactingMessageEvent.newFlags().ifPresent(message::setFlags);
+            impactingMessageEvent.flatMap(ImpactingMessageEvent::newFlags).ifPresent(message::setFlags);
 
-            if (!impactingMessageEvent.wasDeleted()) {
+            if (wasNotDeleted(impactingMessageEvent)) {
```
Comment (mbaechler) [inline mailbox/tools/indexer/src/main/java/org/apache/mailbox/tools/indexer/ReIndexerImpl.java]: please keep positive method naming  `wasDeleted`, it makes the code more easy to understand (wasNotDeleted return true is always a challenge to track for my mind)

File: mailbox/tools/indexer/src/main/java/org/apache/mailbox/tools/indexer/ReIndexerImpl.java (hunk 1 above)
Comment (chibenwa) [reply mailbox/tools/indexer/src/main/java/org/apache/mailbox/tools/indexer/ReIndexerImpl.java]: Latter on this is replaced by an Optional...

I may have messed up an intermediary refactoring step.

**Comment by chibenwa** [comment]
> you don't add a test on a big mailbox?

Checking for not generating OutOfMemory exceptions on top of a memory implementation?

**Comment by chibenwa** [comment]
Welcome to hell

```
[44685273f9d0d7c5d7ac67d7cfd142812a1d108f] [ERROR]   RabbitMQClusterTest$ClusterNodesFailure.connectingToAClusterWithAFailedRabbit » Completion
```

test this please

**Comment by rouazana** [comment]
> Checking for not generating OutOfMemory exceptions on top of a memory implementation?

No, just checking that reindexing a ~100 000 mails mailbox is working.

**Comment by chibenwa** [comment]
> No, just checking that reindexing a ~100 000 mails mailbox is working.

Sorry, I don't see the point. Using memory it was working before as by definition the dataset was fitting in memory... (The number is NOT the issue here)

**Comment by chibenwa** [comment]
Was :green_apple:

**Comment by chibenwa** [comment]
test this please

File: mailbox/tools/indexer/pom.xml [hunk 3]
```diff
@@ -52,6 +63,38 @@
             <type>test-jar</type>
             <scope>test</scope>
         </dependency>
+        <dependency>
+            <groupId>${james.groupId}</groupId>
+            <artifactId>apache-james-mailbox-memory</artifactId>
```
Comment (aduprat) [inline mailbox/tools/indexer/pom.xml]: Should be before `apache-james-mailbox-store`

File: mailbox/tools/indexer/src/test/java/org/apache/mailbox/tools/indexer/CassandraReIndexerImplTest.java [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+import org.junit.After;
+import org.junit.AfterClass;
+import org.junit.Before;
+import org.junit.BeforeClass;
+import org.junit.ClassRule;
+import org.junit.Test;
+
+import com.google.common.base.Strings;
+
+public class CassandraReIndexerImplTest {
```
Comment (aduprat) [inline mailbox/tools/indexer/src/test/java/org/apache/mailbox/tools/indexer/CassandraReIndexerImplTest.java]: Not in JUnit5?

**Comment by chibenwa** [comment]
Merged

**Pull request closed without merging**